akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+  // 16665 + 1024 (room for compressed data) + 1024 (for OpenJDK compatibility)
+  val transportOutBuffer = ByteBuffer.allocate(16665 + 2048)
+  /*
+   * deviating here: chopping multiple input packets into this buffer can lead to
+   * an OVERFLOW signal that also is an UNDERFLOW; avoid unnecessary copying by
+   * increasing this buffer size to host up to two packets
+   */
+  val userOutBuffer = ByteBuffer.allocate(16665 * 2 + 2048)
+  val transportInBuffer = ByteBuffer.allocate(16665 + 2048)
+  val userInBuffer = ByteBuffer.allocate(16665 + 2048)
+
+  val userInChoppingBlock = new ChoppingBlock(UserIn, "UserIn")
+  userInChoppingBlock.prepare(userInBuffer)
+  val transportInChoppingBlock = new ChoppingBlock(TransportIn, "TransportIn")
+  transportInChoppingBlock.prepare(transportInBuffer)
+
+  val engine: SSLEngine = {
+    val e = sslContext.createSSLEngine()
+    e.setUseClientMode(role == StreamTls.Client)
+    e
+  }
+  var currentSession = engine.getSession
+  var currentSessionParameters = firstSession
+  applySessionParameters()
+
+  def applySessionParameters(): Unit = {
+    val csp = currentSessionParameters
+    import csp._
+    enabledCipherSuites foreach (cs ⇒ engine.setEnabledCipherSuites(cs.toArray))
+    enabledProtocols foreach (p ⇒ engine.setEnabledProtocols(p.toArray))
+    clientAuth match {
+      case Some(ClientAuth.None) ⇒ engine.setNeedClientAuth(false)
+      case Some(ClientAuth.Want) ⇒ engine.setWantClientAuth(true)
+      case Some(ClientAuth.Need) ⇒ engine.setNeedClientAuth(true)
+      case None                  ⇒ // do nothing
+    }
+    sslParameters foreach (p ⇒ engine.setSSLParameters(p))
+    engine.beginHandshake()
+    lastHandshakeStatus = engine.getHandshakeStatus
+  }
+
+  def setNewSessionParameters(n: NegotiateNewSession): Unit = {
+    if (tracing) log.debug(s"applying $n")
+    currentSession.invalidate()
+    currentSessionParameters = n
+    applySessionParameters()
+    corkUser = true
+  }
+
+  /*
+   * So here’s the big picture summary: the SSLEngine is the boss, and it can
+   * be in several states. Depending on this state, we may want to react to
+   * different input and output conditions.
+   *
+   *  - normal bidirectional operation (does both outbound and inbound)
+   *  - outbound close initiated, inbound still open
+   *  - inbound close initiated, outbound still open
+   *  - fully closed
+   *
+   * Upon reaching the last state we obviously just shut down. In addition to
+   * these user-data states, the engine may at any point in time also be
+   * handshaking. This is mostly transparent, but it has an influence on the
+   * outbound direction:
+   *
+   *  - if the local user triggered a re-negotiation, cork all user data until
+   *    that is finished
+   *  - if the outbound direction has been closed, trigger outbound readiness
+   *    based upon HandshakeStatus.NEED_WRAP
+   *
+   * These conditions lead to the introduction of a synthetic TransferState
```

Comment:
I was considering the same before, but I got lost in the details at some point :)